main/src/library/SemiGroup.flix

```diff
@@ -31,3 +38,86 @@ instance SemiGroup[Unit] {
 instance SemiGroup[String] {
     pub def combine(x: String, y: String): String = x + y
 }
+
+instance SemiGroup[Option[a]] with SemiGroup[a] {
+  pub def combine(x: Option[a], y: Option[a]): Option[a] = match (x, y) {
+        case (Some(x1), Some(y1)) => Some(SemiGroup.combine(x1, y1))
+        case (a, None) => a
+        case (None, b) => b
+    }
+}
+
+instance SemiGroup[List[a]] {
+  pub def combine(x: List[a], y: List[a]): List[a] = x ::: y
+}
+
+instance SemiGroup[Set[a]] with Order[a] {
+  pub def combine(x: Set[a], y: Set[a]): Set[a] = Set.union(x,y)
+}
+
+instance SemiGroup[Map[k,v]] with Order[k], SemiGroup[v] {
+  pub def combine(x: Map[k,v], y: Map[k,v]): Map[k,v] = Map.unionWith((v1, v2) -> SemiGroup.combine(v1, v2), x, y)
+}
+
+instance SemiGroup[Validation[t, e]] with SemiGroup[t] {
+  pub def combine(x: Validation[t, e], y: Validation[t, e]): Validation[t, e] = match (x, y) {
+        case (Success(x1), Success(y1)) => Success(SemiGroup.combine(x1, y1))
+        case (Failure(es1), Failure(es2)) => Failure(Nel.append(es1, es2))
+        case (Failure(_), _) => x
+        case (_, Failure(_)) => y
```

Comment:
We usually align all the `=>` arrows, i.e. they should line up by indenting the cases.